ml-agents-envs/mlagents/envs/timers.py

```diff
@@ -0,0 +1,172 @@
+    time.sleep(t)
+
+def main():
+    for i in range(3):
+        foo(i + 1)
+    with hierarchical_timer("context"):
+        foo(1)
+
+    print(get_timer_tree())
+
+This would produce a timer tree like
+    (root)
+        "foo"
+        "context"
+            "foo"
+
+The total time and counts are tracked for each block of code; in this example "foo" and "context.foo" are considered
+distinct blocks, and are tracked separately.
+"""
+
+
+class TimerNode:
+    """
+    Represents the time spent in a block of code.
+    """
+
+    __slots__ = ["children", "total", "count"]
+
+    def __init__(self):
+        self.children: Dict[str, TimerNode] = {}
+        self.total: float = 0.0
+        self.count: int = 0
+
+    def get_child(self, name: str) -> "TimerNode":
+        """
+        Get the child node corresponding to the name (and create if it doesn't already exist).
+        """
+        child = self.children.get(name)
+        if child is None:
+            child = TimerNode()
+            self.children[name] = child
+        return child
+
+    def add_time(self, elapsed: float) -> None:
+        """
+        Accumulate the time spent in the node (and increment the count).
+        """
+        self.total += elapsed
+        self.count += 1
+
+
+class TimerStack:
+    """
+    Tracks all the time spent. Users shouldn't use this directly, they should use the contextmanager below to make
+    sure that pushes and pops are already matched.
+    """
+
+    __slots__ = ["root", "stack"]
+
+    def __init__(self):
+        self.root = TimerNode()
+        self.stack = [self.root]
+
+    def push(self, name: str) -> TimerNode:
+        """
+        Called when entering a new block of code that is timed (e.g. with a contextmanager).
+        """
+        current_node: TimerNode = self.stack[-1]
+        next_node = current_node.get_child(name)
+        self.stack.append(next_node)
+        return next_node
+
+    def pop(self) -> None:
+        """
+        Called when exiting a new block of code that is timed (e.g. with a contextmanager).
+        """
+        self.stack.pop()
+
+    def get_timing_tree(self, node: TimerNode = None) -> Dict[str, Any]:
+        """
+        Recursively build a tree of timings, suitable for output/archiving.
+        """
+        if node is None:
+            node = self.root
+
+        res: Dict[str, Any] = {"total": node.total, "count": node.count}
+
+        child_total = 0.0
+        child_list = []
+        for child_name, child_node in node.children.items():
+            child_res: Dict[str, Any] = {
+                "name": child_name,
+                **self.get_timing_tree(child_node),
+            }
+            child_list.append(child_res)
+            child_total += child_res["total"]
+
+        # "self" time is total time minus all time spent on children
+        res["self"] = max(0.0, node.total - child_total)
+        if child_list:
+            res["children"] = child_list
+
+        return res
```

Comment:
It seems that based on this implementation, the root TimerNode is treated as a special node (its total, count is always 0, it doesn't have a name). Since this is a recursive call, should we set a base case so that the root TimerNode also has the corresponding values?

Reply:
I agree, the handling of the root node is a little strange but I'm not sure what the best way to format the output is. A few ways I was thinking about:
* Omit total, self, and count keys from the root
* Just return the list of the root's children
* For the root, set total = program execution time, self = total - children, and count = 1
Sounds like you're suggesting the third one? Any other ways to do it?

Reply:
Yes I'm suggesting the third one. Also I'm suggesting adding a "name" attribute to the TimerNode class, and give the root TimerNode a key value pair as "name":"root". Right now you are adding the "name" attribute to the child TimerNode in the line 104, which was a little bit confusing.

Reply:
Sounds good, I'll do that (and I forgot about adding `"name":"root"`)

I'll add some comments about the node names - since the parent has the name as the key, the node doesn't actually store it.